[28기 박윤국] Add: Mission 구현 완료 + Login 토큰 발급

YunKukPark opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [x] 리뷰 반영
- [x] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 개인 폴더(pages/heewonkang)에 개인이 작업하던 Login, Main 코드 옮기기



|MISSION NUMBER|Page|DESC|구현|
|:--:|:--:|:--|:--:|
|1|Login|사용자 입력 데이터 저장|✅|
|2|Login|로그인 버튼 활성화(validation)|✅|
|3|Main|댓글기능|✅|
|4|Main|댓글 컴포넌트화 + props로 데이터 전달|✅|
|5|Main|map 함수 적용시 key props를 부여하는 이유|✅| 
<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
### Component Tree
<img width="1649" alt="스크린샷 2021-12-23 오전 11 15 54" src="https://user-images.githubusercontent.com/53929065/147177426-10f38d9a-625f-442a-8be0-0b6a1e2c867f.png">

|Name|분리이유|
|:--:|--|
|MainLeft / Right| Main 구조화 |
|Story, Feed| MainLeft 구조화|
|userProfile, Recommadation|MainRight 구조화|
|FeedDesc, CommentInput| Feed 구조화 <br/> 이미지부분, 설명 및 댓글창, 댓글 입력란 |
|...List|map을 이용하여 반복 출력을 위함|

<br/>

|Page|구현사항|구현이미지|추가 desc|
|:--:|:--|:--:|--|
|Login|로그인 사용자 입력 데이터 저장|![로그인 Validate](https://images.velog.io/images/yunkuk/post/c2b10228-af61-4feb-87b0-e699cb8d00b5/%E1%84%89%E1%85%B3%E1%84%8F%E1%85%B3%E1%84%85%E1%85%B5%E1%86%AB%E1%84%89%E1%85%A3%E1%86%BA%202021-12-11%20%E1%84%8B%E1%85%A9%E1%84%92%E1%85%AE%204.54.40.png)|로그인 입력 데이터 유효성 검사<br/>1. 입력창에 변화가 일어날 때 마다 현재 입력값이 있는지 확인하는 함수 실행<br/>2. 입력값 검사 함수의 반환값(boolean)에 따라 버튼의 배경색과 disabled 속성이 변화한다.<br/>3. 아이디에 `@` 가 없거나 비밀번호가 5글자 이상이 아니면 로그인이 되지 않는다.|
|Main|SetLayout 함수 구현|![스크린샷 2021-12-10 오후 5 52 15](https://user-images.githubusercontent.com/53929065/145545642-ad6c1c80-b3cb-4140-a52e-f1b2128ccfb9.png)|현재 사용자의 window가 `resize`되거나 `load` 되어도 **오른쪽의 `aside`가 같은 공간에 붙어있을 수 있도록 `right` 값을 부여**해주는 함수<br/> - `useEffect`를 이용하여 `mount`될 때 `handleLayout`이 작동하고, `unmount`될 때 `clean-up`된다.|
|Main|feedLike 함수 구현|![ezgif-4-9c876de549b6](https://user-images.githubusercontent.com/53929065/145546448-e4fc5a41-eb5d-4d2a-be7b-20225ee5240b.gif)|feed이미지 더블클릭, 좋아요 클릭 시 좋아요 갯수가 증가하며, 좋아요 icon이 토글링 <br/> - setState를 활용하여 Like가 되었을 때 클래스 부여 <br/> - setState를 활용하여 좋아요 개수 handle|
|Main|Comment Submit 구현|![ezgif-4-794fbabb0cee](https://user-images.githubusercontent.com/53929065/145546940-a90984a0-7073-4825-8ffa-5a1b76c15f12.gif)|```setFeedComments([newComment, ...feedComments])```|
|Main|Comment Delete 구현|![](https://images.velog.io/images/yunkuk/post/57bf656e-cf11-451d-81ec-ce8610066ec0/ezgif-2-f4e96c3da0d1.gif)|```[...feedComments].filter(feedComment => feedComment.id !== comment.id)```|
|Main|user mymenu Dropdown 구현|![](https://images.velog.io/images/yunkuk/post/cc2fb751-6537-4ec2-ac9f-137a61b978ab/ezgif-2-433a12f085a4.gif)|dropdown 구현 <br/>- 실제 인스타 페이지처럼 프로필 사진 클릭 시 메뉴 박스가 생성될 수 있도록 구현<br/> - 메뉴 박스 바깥 영역 클릭 시에는 다시 닫히도록 구현|
|Main|Mock Data 만들기 |![](https://images.velog.io/images/yunkuk/post/298a4ed3-6f1b-4601-961f-cdfda8a32d81/%E1%84%89%E1%85%B3%E1%84%8F%E1%85%B3%E1%84%85%E1%85%B5%E1%86%AB%E1%84%89%E1%85%A3%E1%86%BA%202021-12-21%20%E1%84%8B%E1%85%A9%E1%84%92%E1%85%AE%203.38.22.png) | 목업용 json 파일 만든 후 Fetch|
|Main|토큰 확인 로직 구현 |<img width="879" alt="스크린샷 2021-12-23 오전 11 18 05" src="https://user-images.githubusercontent.com/53929065/147177584-38385fc3-210d-49ba-ba55-4f3c57f766ea.png"> |토큰이 확인되지 않은 유저 로그인 페이지로 추방|


<br />

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- 협업의 첫 과정을 천천히 따라갈 수 있었습니다.
- useState를 활용하여 값을 조건에 따라 동적으로 변경될 수 있도록 이벤트를 주었습니다.
- useEffect를 활용하여 component가 Mount되었을 때 데이터를 패칭 할 수 있었습니다.
- useEffect를 활용하여 이벤트를 주고, 컴포넌트가 사라졌을 때 이벤트를 clean up 하는 과정을 밟아봤습니다.
- props로 부모의 데이터를 자식에게 부여하고, state끌어올리기를 사용하여 자식의 변경데이터를 상위 state에게 전달하는 방법을 알게 되었습니다.
- 리액트 Infinite 로드를 경험해보며, 리액트 생명주기, 왜 무한로드가 되는지 경험했습니다.
- back 과의 두근두근 첫 통신으로 back에서 보내준 api를 front에서 가공 하는 것을 경험했습니다. 


<br />

## :: 기타 질문 및 특이 사항
- Input 을 핸들링 할 때 input에 한글자씩 채워질 때 마다 재 렌더링이 되는데, 이러한 Rendering 을 Reduce 하고 싶은데 어떻게 하면 좋을까요?

YunKukPark committed:
[Add] 기존 데이터 마이그레이션

YunKukPark committed:
[Refactor] Separate Story.jsx to Story & StoryList
- 컴포넌트 분해 작업

YunKukPark committed:
Refactor: WOP Feed Mock Data & Separating

YunKukPark committed:
Fix: & 미사용으로 인한 스타일 적용 버그 수정

YunKukPark committed:
Refactor: 회원님을 위한 추천 컴포넌트 분리

YunKukPark committed:
Refactor: Main Feed 관련 컴포넌트 분리

YunKukPark committed:
Fix: Component Mount시 handleLayout 함수가 작동하지 않는 점 Fix

YunKukPark committed:
Add: Dropdown 구현 중이나 버그 발견

YunKukPark committed:
Add: Mission3, 4)  댓글 추가, 삭제 기능 구현
- Mission 3 댓글 기능 구현
- Mission 4 댓글 컴포넌트화 + Props로 데이터 전달 구현

YunKukPark committed:
Add: comment Mock data 추가

YunKukPark committed:
Fix: simple fix for a non-critical issue

YunKukPark committed:
Fix/Add: simple Fix(한글인코딩문제), 댓글 3개 이상 arcodian 주기
- fix: 한글일 경우 댓글이 2번 등록되는 경우가 있었는데 keyDown => keyPress 로 변경 해줌
- fix: input 창이 비어있을 때 submit 되지 않게 prevent
- refactor: setInputInit 이라는 함수로 Input Cleaner 따로 뺌
= Add : 댓글 3개 이상 일 시 댓글 모두보기 창 활성화
= Add : 댓글 모두보기 창은 아코디언 형식으로 댓글을 다 보여줬다 숨겼다함

YunKukPark committed:
Fix: Clean up useEffect추가

YunKukPark committed:
Add: Mock Data 추가 후 데이터 패칭

YunKukPark committed:
Modify: misspell sider 오류 수정

YunKukPark committed:
Modify: misspell sider 폴더명 오류 수정

YunKukPark committed:
Modify: misspell sider 파일 내 컴포넌트 misspell 수정

seankim1111 reviewed (commented):
고생하셨습니다 윤국님! 코드가 깔끔하네요🙂

KimJeongHyun commented on src/pages/yunkukPark/Login/Login.js:
```diff
@@ -1,3 +1,102 @@
+    </div>
+  );
+};
+
```
별도의 useEffect 등의 동작 없이 input handling을 통해 button disabled 유무까지 흐름을 이은 부분이 너무 좋네요 특히 저는 state로 어떠한 작업을 컨트롤하는 데에 신경을 쓰는 편인데, style도 같이 제어하는 부분 많은 참고됩니다  👍👍

KimJeongHyun commented on src/pages/yunkukPark/Main/Feed/CommentList.jsx:
```diff
@@ -0,0 +1,43 @@
+import React, { useState } from 'react';
+
+const CommentList = props => {
```
얼핏 댓글을 전부 wrapping하는 컴포넌트로 생각했습니다 😅

YunKukPark replied:
저만의 컨벤션이 무언가를 반복하면 ...List로 관리하자 였었거든요 ㅋㅋㅋ..
Wrapping 하는 컴포넌트 였다면 ...Items 로 명명 해줘도 좋을 것 같군요!
좋은 지적 감사합니다 💪🏻 🙌

KimJeongHyun commented on src/pages/yunkukPark/Main/Feed/CommentList.jsx:
```diff
@@ -0,0 +1,43 @@
+    comment,
+    comment: { userName, content },
+  } = props;
+
```
![image](https://user-images.githubusercontent.com/23470125/146889002-a6d10c9f-da64-44eb-ab51-cae7ea1b8b87.png)

저는 props 부분에 바로 구조분해를 진행했는데, 윤국님처럼 안 쪽에 넣어주는 것도 변수 기억의 측면에서 좋은 것 같습니다!

KimJeongHyun commented on src/pages/yunkukPark/Main/Feed/CommentList.jsx:
```diff
@@ -0,0 +1,43 @@
+
+  const [isCommentLike, setIsCommentLike] = useState(false);
+  const [likeClass, normalClass] = ['fas is-liked', 'far'];
+
```
likeClass, normalClass를 분리하신 이유가 있을까요??

YunKukPark replied:
`likeClass`는 좋아요(`isLike State가 True`)를 눌렀을 때의 하트 이고, 
`normalClass` 는 일반 상태의 하트 입니다. 

`isLike` State 에 따라 className 을 다르게 주고 싶어서 따로 분리해서 사용했습니다. 

그리고 왜 굳이 구조분해할당으로 했냐면  
```jsx
const likeClass = 'fas is-liked'
const normalClass = "far"
```
이렇게 두 줄로 쓰는 것 보다
한줄에 state 에 따라 종속적으로 움직이는 변수이다 라는걸 보는 차원에서 한눈에 알아볼 수 있게 하고 싶어서 그랬습니다!!

KimJeongHyun commented on src/pages/yunkukPark/Main/Feed/FeedDesc.jsx:
```diff
@@ -0,0 +1,48 @@
+import CommentList from './CommentList';
+
+const FeedDesc = props => {
+  const { userName, content, comments, onDeleteButtonClick } = props;
```
혹시 FeedDescription이 아닌 FeedFooter는 어떨까요??

view에 feed-desc 라는 영역과 comment-list 영역이 분리되어 렌더링되고 있는데
그에 반해 얼핏 이 컴포넌트는 피드의 내용만 담겨져있다! 라고 생각될 수도 있을 것 같습니다

YunKukPark replied:
이부분 되게 좋은 것 같습니다! 
저도 컴포넌트 흐름 관리 할 때 Description 이라고 해서 조금 애매 했었거든요! 
내일 적용하고 답 남겨놓도록 하겠습니다. 

- [ ] FeedFooter로 수정

YunKukPark replied:
✅ Done in [Commit URL](https://github.com/wecode-bootcamp-korea/28-React-Westagram-1/pull/4/commits/dcf1b27b8f7b62f3ec8be0acb3392a3e10f5dc04)

KimJeongHyun commented on src/pages/yunkukPark/Main/MainLeft/MainLeft.jsx:
```diff
@@ -0,0 +1,26 @@
+  const [feedContents, setfeedContents] = useState([]);
+
+  useEffect(() => {
+    fetch(`${DATA_URL}/feed.json`)
```
유지보수적인 측면에서 URL 상수로 따로 빼신 것 좋습니다!!

KimJeongHyun commented on src/pages/yunkukPark/Main/MainRight/MainRight.jsx:
```diff
@@ -0,0 +1,36 @@
+      mainLeftXoffset + mainLeftWidth + mainLeftMarginRight;
+    setMainRightLeft(mainRightXoffset);
+  };
+
```
저번에 구두로 논의했던 내용에서 다뤄졌던 부분인 것 같습니다.

![image](https://user-images.githubusercontent.com/23470125/146890803-b0839319-b06f-4106-b8b3-45f5021a5035.png)

제 css 참고차 올려드립니다! 세세한 값들은 제외하고 display, position을 큰 틀로 이렇게 했구나~ 라고 생각해주시면 됩니다.

잘 갈무리해보면 offset 속성 사용을 줄일 수 있을 것 같습니다!

KH2Wone commented on src/pages/yunkukPark/Login/Login.js:
```diff
@@ -1,3 +1,102 @@
+  const [password, setPassword] = useState('');
+  const [buttonSwitch, setButtonSwitch] = useState(true);
+
+  const handleInput = event => {
+    const {
+      target: { name, value },
+    } = event;
+    if (name === 'user-id') setId(value);
+    if (name === 'user-password') setPassword(value);
+
+    handleButton(id, password);
+  };
```
아니 이런 방법이... 코드가 정말 깔끔합니다 💨

YunKukPark replied:
저도 맨 처음엔 id 와 password 를 따로 분리 하였으나 중복된 코드가 있어서 이걸 따로 묶어주면 좋겠다! 라고 생각했습니다. 
그래서 생각한게 `input`의 `name` 속성을 활용해보자 해서 다음과 같이 코드를 작성했습니다! 💪🏻

KH2Wone commented on src/pages/yunkukPark/Login/Login.js:
```diff
@@ -1,3 +1,102 @@
+    return validateId() && validatePassword();
+  };
+
+  const validateId = () => {
+    if (id.indexOf('@') === -1) {
+      alert('님 아이디에 @ 안씀');
+      return false;
+    }
+    return true;
+  };
+
+  const validatePassword = () => {
+    if (password.length > 5) {
+      return true;
+    }
+    if (password.length <= 5) {
+      alert('님 비밀번호 5글자 이상 아님?');
+      return false;
+    }
+  };
```
저는 `&&`로 묶는것까지 생각하고 이걸 어떻게 아이디와 비밀번호를 다 연결하지 고민했는데요...
유효성 검사를 각각 `id`, `password`로 나눠서 관리하시는거 보고 많이 배워갑니다 ,,,,, ✨
이러면 `navigate`까지 다 컨트롤 할 수 있네요!!

YunKukPark replied:
넵! 각각의 `validate` 는 `boolean` 값을 `return` 해줍니다
> 이렇게 하였을 시 1함수 1동작의 원리를 잘 적용할 수 있더라구요!! 🙌

KH2Wone commented on src/pages/yunkukPark/Main/Main.scss:
```diff
@@ -0,0 +1,606 @@
+      border-radius: 50%;
+    }
+
+    &-content {
```
전체적인 `scss` 파일을 보면 `&` 활용 너무 잘하셨네요. 게을렀던 저 반성하게 됩니다 ,,

YunKukPark replied:
mixin 과 constant 말고 비슷하지만 조금씩 다른 컴포넌트 (ex. `user-avatar`)도 
큰 틀을 만들어 놓고 extends 받아서 쓰지 못해서 아쉽습니다. 
1차 프로젝트 때는 꼭 거기까지 경험 해보고 싶어요! 🙌

KH2Wone reviewed (commented):
윤국님 고생 많으셨습니다 !! 👍🏻
구현한 것도 많으시고 코드 가독성도 좋아요.
보고 많이 배웠습니다~

YunKukPark committed:
Modify: 리뷰반영 feedDesc => feedFooter 컴포넌트 이름 변경

YunKukPark committed:
Add: 백엔드와 두근두근 첫통신
- 회원가입 페이지 추가
- 회원 가입 진행 (POST)
- 로그인 진행 (POST)
- 유저정보 토큰으로 계속 가지고 있기
- 유저정보 통신

YunKukPark committed:
Add: 메인페이지에 접근 시 토큰이 없으면 추방

YunKukPark committed:
Fix: simple fix variable  never used Issue

YunKukPark committed:
Fix: css 중복 선언 되는 부분 제거

YunKukPark committed:
Refactor: FooterList 컴포넌트 분리
- footer list 상수데이터 컴포넌트분리
- 반복 내용 map으로 돌림

YunKukPark committed:
Fix: import 문 변경

YunKukPark committed:
Fix: Recommendation 에 종속된 Footer  컴포넌트 분리
- Recommendation 안에 Footer가 종속되어 필요없는 렌더가 발생되는 이슈 발견
- 종속관계일 필요가 없어 MainRight 로 Footer 컴포넌트 이동

YunKukPark committed:
Refactor: 로그인 페이지 리팩토링
- 계산된 속성명을 이용한 inputHandler 함수 합치기
- API 받아오는 부분 config.js 로 따로 관리
- loginbutton on/off 는 validate를 통해 해결
- 의미에 맞는 함수명 작명

YunKukPark committed:
Refactor: MainContents 부분 리팩토링
- Main.js : 유저 확인 절차 코드 간소화
- MainLeft, Right : 이름 변경 (MainContents, Aside)
- CommentInput : commentLenght 를 직접 받아오지 않고 comment를 받아서 안에서 처리
- CommentInput : misspell 수정
- Feed.jsx : 좋아요 관련 setLikeInfo 간소화
- 기타 : misspell 수정

YunKukPark committed:
Fix: misspell 해결

YunKukPark committed:
Refactor: Signup page, CommentList, NavMyMenu  리팩토링
